### PR TITLE
Feature/6/ui mode info

### PR DESCRIPTION
**Pull request by chfsx**
@klees and @Amstutz this is a call for ultra fast lane review :D as discussed last friday, ILIAS may need a ModeInfo Component:

image

The following PR contains the Interface-Chynge on StandardPage and for a new component ModeInfo.

### CODE REVIEW

**Comment by chfsx** [comment]
Changes:
- New Methods on StandardPage: https://github.com/ILIAS-eLearning/ILIAS/pull/2328/files#diff-cd550c325169e2bd54eb3649a292c6b2R92
- Description of new component ModeInfo: https://github.com/ILIAS-eLearning/ILIAS/pull/2328/files#diff-19b7a6034a22226f59300948731bc31fR322
- Interface of ModeInfo: https://github.com/ILIAS-eLearning/ILIAS/pull/2328/files#diff-4c70382d72d36359cee2f2476400ca8cR15

File: src/UI/Component/MainControls/Factory.php [hunk 1]
```diff
@@ -313,4 +315,38 @@ public function slate() : Slate\Factory;
...
+
+
+    /**
+     * ---
+     * description:
+     *   purpose: >
+     *     The ModeInfo is a section of the page that informs about a certain
+     *     (application wide) mode in which the user is viewing ILIAS.
+     *     Such modes can occur e.g. in the preview as a member or in LTI.
+     *     The Mode Info MUST allow the user to leave the mode.
```
Comment (klees) [inline src/UI/Component/MainControls/Factory.php]: Please move to `rules.interaction`.

File: src/UI/Component/MainControls/Factory.php [hunk 2]
```diff
@@ -313,4 +315,38 @@ public function slate() : Slate\Factory;
...
+     *     Such a mode can be leaved by the user by clicking the Close Button.
+     *
+     *   composition: >
+     *     The Mode Info contains a title explaining the state.
+     *     The ModeInfo MUST also contain a Close Button to leave the
```
Comment (klees) [inline src/UI/Component/MainControls/Factory.php]: Please RFC 2119-language in the rules section and less formal language in the description.

File: src/UI/Component/MainControls/Factory.php [hunk 3]
```diff
@@ -313,4 +315,38 @@ public function slate() : Slate\Factory;
...
+     *
+     * rules:
+     *   usage:
+     *     1: The Mode Info is unique for the page - there MUST be not more than one.
+     * ----
+     *
+     * @param string $title
+     * @param URI    $close_action
+     *
+     * @return ModeInfo
```
Comment (klees) [inline src/UI/Component/MainControls/Factory.php]: Please remove the docstrings. They just duplicate information from the (executable) type hints.

**Review by klees — APPROVED** [review]
Hi @chfsx!

LGTM, left some very minor change requests.

Thanks for this!

**Review by Amstutz — CHANGES_REQUESTED** [review]
Hi @chfsx 

Where do you take the time from? Your output is insane, congrats!

Other than the comments of @klees I would like to have 2 accessability rules on the Mode Info, so plz add:
- [x] Accessabilty Role: Since for blind People, this Info is IMO critical, plz add a rule requiring to add the proper WAI-Role to this element. Since I am not expert on this either, we best ask the JF for a decision on this here.
- [x] ARIA Label: Finde a proper label for this. E.g. something like Mode Info.

As mentioned, I would like to discuss this in the upcoming JF and hope that it will be accepted. Thx a lot.

**Comment by chfsx** [comment]
Hi @Amstutz 
Thanks for those additions, I'll tray to tackle them. but concerning the WAI-Role, I do not find any matching: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles. Best I can find is AlertRole (which I'll propose this afternoon). Any better Idea?

**Comment by Amstutz** [comment]
No, that is why, I moved the question along forward to the JF. I know, that there are people hopefully attending with more expertise then I have (e.g. @atoedt ). If we do not find anything, we have to drop this proposal of mine I am afraid.

File: src/UI/Component/MainControls/Factory.php [hunk 4]
```diff
@@ -313,4 +315,36 @@ public function slate() : Slate\Factory;
...
+     *
+     * rules:
+     *   usage:
+     *     1: The Mode Info is unique for the page - there MUST be not more than one.
+     *   interaction:
+     *     1: The Mode Info MUST allow the user to leave the mode.
+     *   accessibility:
+     *     1: The Mode Info informs about an important circumstance, which must be
+     *     recognizable in particular also for persons with a handicap. Therefore a
+     *     role="alert" MUST be implemented.
```
Comment (matthiaskunkel) [inline src/UI/Component/MainControls/Factory.php]: We remove the last sentence in 'accessibility' to keep the discussion open which way is the best to make this UI component accessible.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 18 NOV 2019 : We highly appreciate the suggested PR and accept it for ILIAS 6. We removed the last sentence in 'Accessibility' to keep the discussion open which way is the best to make this UI component accessible.

**Comment by chfsx** [comment]
see #2359 which includes this PR as well

**Pull request closed without merging**